airbyte-cdk/python/airbyte_cdk/sources/declarative/transformations/add_fields.py

```diff
@@ -110,10 +110,11 @@ def transform(
         stream_state: Optional[StreamState] = None,
         stream_slice: Optional[StreamSlice] = None,
     ) -> Record:
-        kwargs = {"record": record, "stream_state": stream_state, "stream_slice": stream_slice}
-        for parsed_field in self._parsed_fields:
-            value = parsed_field.value.eval(config, **kwargs)
-            dpath.util.new(record, parsed_field.path, value)
+        if isinstance(record, Mapping):
```

Comment:
This problem starts here
https://github.com/airbytehq/airbyte/blob/master/airbyte-cdk/python/airbyte_cdk/sources/declarative/retrievers/simple_retriever.py#L412

in debug mode, we emit to stream of records also `AirbyteMessage[type=LOG]`

I am not the first who added this check

[here1](https://github.com/airbytehq/airbyte/blob/master/airbyte-cdk/python/airbyte_cdk/sources/declarative/retrievers/simple_retriever.py#L377)
[here2](https://github.com/airbytehq/airbyte/blob/master/airbyte-cdk/python/airbyte_cdk/sources/declarative/retrievers/simple_retriever.py#L382)
[here3](https://github.com/airbytehq/airbyte/blob/master/airbyte-cdk/python/airbyte_cdk/sources/declarative/partition_routers/substream_partition_router.py#L150)

Reply:
This is already a discussion with had with @girarda. The conclusion was that it's a bad smell to have to `isintance` before doing something. It will likely be very hard to maintain in the long run because it means that if we had another "type" of record, we need to go find all the places where there is a `isinstance`. As we can see here, we will probably miss some. 

So, what is the solution? If we really consider debug message as records, I don't see any other solution than to have interaction on the record be done in the record level. For the specific example here, it would look like:
```
class Record:
  def transform(self, transformer):
    raise NotImplementedError()
<...>

class SourceRecord:
  def transform(self, transformer):
    transformer.transform(<...>)
    
class DebugRecord:
  def transform(self, transformer):
    # not applicable for debug message
```
Note: I don't like the naming here but this is just for showing what it could look like. 

If we don't consider these as records, it needs to be reflected in our code.

Now, this is clearly out of scope for this story so I approve Serhii's change. However, we should probably define a plan, agree on it and split it in chunk so that we are reducing the impact of such design. We can track this on this issue : https://github.com/airbytehq/airbyte/issues/23359 In the meanwhile, I think Serhii's change is the way to go

Reply:
100% agree that we need to improve design. I would however prefer we do this after we're done with the interface between the connector builder and the CDK to avoid throwaway work.

Reply:
fix is in https://github.com/airbytehq/airbyte/pull/24827